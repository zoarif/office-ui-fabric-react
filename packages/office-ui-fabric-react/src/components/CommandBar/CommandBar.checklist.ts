import { ChecklistStatus } from '../../demo/ComponentStatus/ComponentStatus.types';

export const CommandBarStatus = {
  keyboardAccessibilitySupport: ChecklistStatus.unknown,
  markupSupport: ChecklistStatus.unknown,
  highContrastSupport: ChecklistStatus.pass,
  rtlSupport: ChecklistStatus.pass,
  testCoverage: ChecklistStatus.good
};