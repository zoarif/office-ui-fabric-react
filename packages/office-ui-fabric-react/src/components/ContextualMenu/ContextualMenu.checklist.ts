import { ChecklistStatus } from '../../demo/ComponentStatus/ComponentStatus.types';

export const ContextualMenuStatus = {
  keyboardAccessibilitySupport: ChecklistStatus.unknown,
  markupSupport: ChecklistStatus.unknown,
  highContrastSupport: ChecklistStatus.fail,
  rtlSupport: ChecklistStatus.fail,
  testCoverage: ChecklistStatus.good
};