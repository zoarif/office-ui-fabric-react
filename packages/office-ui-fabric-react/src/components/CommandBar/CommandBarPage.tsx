import * as React from 'react';
import {
  ExampleCard,
  IComponentDemoPageProps,
  ComponentPage,
  PropertiesTableSet,
  PageMarkdown,
} from '@uifabric/example-app-base';
import { items, overflowItems, farItems } from './examples/data';
import { CommandBarBasicExample } from './examples/CommandBar.Basic.Example';
import { CommandBarNonFocusableItemsExample } from './examples/CommandBar.NonFocusable.Example';
import { CommandBarCustomizationExample } from './examples/CommandBar.Customization.Example';
import { ComponentStatus } from '../../demo/ComponentStatus/ComponentStatus';
import { CommandBarStatus } from './CommandBar.checklist';

const CommandBarBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/CommandBar/examples/CommandBar.Basic.Example.tsx') as string;
const CommandBarNoFocusableItemsExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/CommandBar/examples/CommandBar.NonFocusable.Example.tsx') as string;
const CommandBarCustomizationExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/CommandBar/examples/CommandBar.Customization.Example.tsx') as string;

export class CommandBarPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    const cmdBarParamsTextAndIcons: any = { items, overflowItems, farItems };

    return (
      <ComponentPage
        title='CommandBar'
        componentName='CommandBarExample'
        componentUrl='https://github.com/OfficeDev/office-ui-fabric-react/tree/master/packages/office-ui-fabric-react/src/components/CommandBar'
        exampleCards={
          <div>
            <ExampleCard title='CommandBar with search box and overflowing menu items' code={ CommandBarBasicExampleCode }>
              <CommandBarBasicExample { ...cmdBarParamsTextAndIcons } />
            </ExampleCard>
            <ExampleCard title='CommandBar with non-focusable items' code={ CommandBarNoFocusableItemsExampleCode }>
              <CommandBarNonFocusableItemsExample />
            </ExampleCard>
            <ExampleCard title='CommandBar with customized rendered items' code={ CommandBarCustomizationExampleCode }>
              <CommandBarCustomizationExample />
            </ExampleCard>
          </div>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/CommandBar/CommandBar.types.ts'),
              require<string>('!raw-loader!office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.types.ts')
            ] }
          />
        }
        overview={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/CommandBar/docs/CommandBarOverview.md') }
          </PageMarkdown>
        }
        bestPractices={
          <div />
        }
        dos={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/CommandBar/docs/CommandBarDos.md') }
          </PageMarkdown>
        }
        donts={
          <PageMarkdown>
            { require<string>('!raw-loader!office-ui-fabric-react/src/components/CommandBar/docs/CommandBarDonts.md') }
          </PageMarkdown>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
        componentStatus={
          <ComponentStatus
            { ...CommandBarStatus }
          />
        }
      />
    );
  }

}
