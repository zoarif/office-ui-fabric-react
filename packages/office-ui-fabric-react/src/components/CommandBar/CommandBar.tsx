import * as React from 'react';
import {
  BaseComponent,
  buttonProperties,
  anchorProperties,
  css,
  divProperties,
  getId,
  getNativeProps
} from '../../Utilities';
import { ICommandBar, ICommandBarProps, ICommandBarItemProps } from './CommandBar.types';
import { IFocusZone, FocusZone, FocusZoneDirection } from '../../FocusZone';
import { ContextualMenu, IContextualMenuProps, IContextualMenuItem } from '../../ContextualMenu';
import { hasSubmenu } from '../../utilities/contextualMenu/index';
import { DirectionalHint } from '../../common/DirectionalHint';
import {
  Icon,
  IIconProps
} from '../../Icon';
import { FontClassNames } from '../../Styling';
import { TooltipHost } from '../../Tooltip';
import * as stylesImport from './CommandBar.scss';
import { createRef } from '../../Utilities';

const styles: any = stylesImport;

const OVERFLOW_KEY = 'overflow';
const OVERFLOW_WIDTH = 41.5;

export interface ICommandBarState {
  renderedItems?: IContextualMenuItem[];
  renderedOverflowItems?: IContextualMenuItem[];
  expandedMenuItemKey?: string;
  expandedMenuId?: string;
  contextualMenuProps?: IContextualMenuProps;
  contextualMenuTarget?: HTMLElement;
  renderedFarItems?: IContextualMenuItem[];
}

export class CommandBar extends BaseComponent<ICommandBarProps, ICommandBarState> implements ICommandBar {
  public static defaultProps: ICommandBarProps = {
    items: [],
    overflowItems: [],
    farItems: []
  };

  public refs: {
    [key: string]: React.ReactInstance;
  };

  private _searchSurface = createRef<HTMLDivElement>();
  private _commandSurface = createRef<HTMLDivElement>();
  private _commandBarRegion = createRef<HTMLDivElement>();
  private _farCommandSurface = createRef<HTMLDivElement>();
  private _focusZone = createRef<IFocusZone>();
  private _overflow = createRef<HTMLDivElement>();

  private _id: string;
  private _overflowWidth: number;
  private _commandItemWidths: { [key: string]: number } | null;

  constructor(props: ICommandBarProps) {
    super(props);

    this.state = this._getStateFromProps(props);

    this._id = getId('CommandBar');
  }

  public componentDidMount(): void {
    // Asynchronously update command bar layout to eliminate forced synchronous reflow
    this._asyncMeasure();
    this._events.on(window, 'resize', this._updateRenderedItems);
  }

  public componentWillReceiveProps(nextProps: ICommandBarProps): void {
    this.setState(this._getStateFromProps(nextProps));
    this._commandItemWidths = null;
  }

  public componentDidUpdate(prevProps: ICommandBarProps, prevStates: ICommandBarState) {
    if (!this._commandItemWidths) {
      // Asynchronously update command bar layout to eliminate forced synchronous reflow
      this._asyncMeasure();
    }
  }

  public render(): JSX.Element {
    const { isSearchBoxVisible, searchPlaceholderText, className } = this.props;
    const { renderedItems, contextualMenuProps, expandedMenuItemKey, expandedMenuId, renderedOverflowItems, contextualMenuTarget, renderedFarItems } = this.state;
    let searchBox;

    if (isSearchBoxVisible) {
      searchBox = (
        <div className={ css('ms-CommandBarSearch', styles.search) } ref={ this._searchSurface }>
          <input className={ css('ms-CommandBarSearch-input', styles.searchInput) } type='text' placeholder={ searchPlaceholderText } />
          <div
            className={ css(
              'ms-CommandBarSearch-iconWrapper ms-CommandBarSearch-iconSearchWrapper',
              styles.searchIconWrapper, styles.searchIconSearchWrapper) }
          >
            { <Icon iconName={ 'Search' } /> }
          </div>
          <div
            className={ css(
              'ms-CommandBarSearch-iconWrapper ms-CommandBarSearch-iconClearWrapper',
              FontClassNames.small,
              styles.searchIconWrapper,
              styles.searchIconClearWrapper
            ) }
          >
            <Icon iconName='cancel' />
          </div>
        </div>
      );
    }
    // Total # of menu items is regular items + far items + 1 for the ellipsis, if necessary
    const setSize = renderedItems!.length + renderedFarItems!.length + (renderedOverflowItems && renderedOverflowItems.length > 0 ? 1 : 0);
    let posInSet = 1;

    return (
      <div className={ css('ms-CommandBar', styles.root, className) } ref={ this._commandBarRegion }>
        { searchBox }
        <FocusZone componentRef={ this._focusZone } className={ styles.container } direction={ FocusZoneDirection.horizontal } role='menubar' >
          <div className={ css('ms-CommandBar-primaryCommands', styles.primaryCommands) } ref={ this._commandSurface }>
            { renderedItems!.map(item => (
              this._renderItemInCommandBar(item, posInSet++, setSize, expandedMenuItemKey!)
            )).concat((renderedOverflowItems && renderedOverflowItems.length) ? [
              <div className={ css('ms-CommandBarItem', styles.item) } key={ OVERFLOW_KEY } ref={ this._overflow }>
                <button
                  type='button'
                  id={ this._id + OVERFLOW_KEY }
                  className={ css('ms-CommandBarItem-link', styles.itemLink, {
                    ['is-expanded ' + styles.itemLinkIsExpanded]: (expandedMenuItemKey === OVERFLOW_KEY)
                  }) }
                  onClick={ this._onOverflowClick }
                  role='menuitem'
                  aria-expanded={ this.state.expandedMenuItemKey === OVERFLOW_KEY }
                  aria-label={ this.props.elipisisAriaLabel || '' }
                  aria-haspopup={ true }
                  aria-setsize={ setSize }
                  aria-posinset={ posInSet++ }
                  data-automation-id='commandBarOverflow'
                >
                  <Icon className={ css('ms-CommandBarItem-overflow', styles.itemOverflow) } iconName='more' />
                </button>
              </div>
            ] : []) }
          </div>
          <div className={ css('ms-CommandBar-sideCommands', styles.sideCommands) } ref={ this._farCommandSurface }>
            { renderedFarItems!.map(item => (
              this._renderItemInCommandBar(item, posInSet++, setSize, expandedMenuItemKey!, true)
            )) }
          </div>
        </FocusZone>
        { (contextualMenuProps) ?
          (<ContextualMenu
            className={ css('ms-CommandBar-menuHost') }
            directionalHint={ DirectionalHint.bottomAutoEdge }
            { ...contextualMenuProps }
            target={ contextualMenuTarget }
            labelElementId={ expandedMenuId }
            onDismiss={ this._onContextMenuDismiss }
          />
          ) : (null)
        }
      </div >
    );
  }

  public focus() {
    const { current: focusZone } = this._focusZone;
    focusZone && focusZone.focus();
  }

  private _renderItemInCommandBar(item: ICommandBarItemProps, posInSet: number, setSize: number, expandedMenuItemKey: string, isFarItem?: boolean): JSX.Element {
    if (item.onRender) {
      return (
        <div className={ css('ms-CommandBarItem', styles.item, item.className) } key={ item.key } ref={ item.key }>
          { item.onRender(item, this._onContextMenuDismiss) }
        </div>
      );
    }

    const itemKey = item.key || String(posInSet);
    const isLink = item.onClick || hasSubmenu(item);
    const className = css(
      isLink ? ('ms-CommandBarItem-link ' + styles.itemLink) : ('ms-CommandBarItem-text ' + styles.itemText),
      !item.name && ('ms-CommandBarItem--noName ' + styles.itemLinkIsNoName),
      (expandedMenuItemKey === item.key) && ('is-expanded ' + styles.itemLinkIsExpanded),
      item.inactive ? styles.inactive : ''
    );

    let tooltipContent = '';

    if (item.title) {
      tooltipContent = item.title;
    }

    const hasIcon = !!item.icon || !!item.iconProps;
    const isNameVisible = !!item.name && !item.iconOnly;
    const ariaLabel = item.ariaLabel || (item.iconOnly ? item.name : undefined);

    let command: React.ReactNode;
    if (item.href) {
      // Allow the disabled property on anchor elements for commandbar
      command = (
        <a
          { ...getNativeProps(item, anchorProperties.concat(['disabled'])) }
          id={ this._id + item.key }
          className={ className }
          href={ item.disabled ? undefined : item.href }
          onClick={ this._onItemClick(item) }
          title={ '' }
          aria-disabled={ item.inactive }
          data-command-key={ itemKey }
          aria-haspopup={ hasSubmenu(item) }
          role='menuitem'
          aria-label={ ariaLabel }
          aria-setsize={ setSize }
          aria-posinset={ posInSet }
        >
          { (hasIcon) ? this._renderIcon(item) : (null) }
          { isNameVisible && (
            <span
              className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
            >
              { item.name }
            </span>
          ) }
        </a>
      );
    } else if (isLink) {
      command = (
        <button
          { ...getNativeProps(item, buttonProperties) }
          id={ this._id + item.key }
          className={ className }
          onClick={ this._onItemClick(item) }
          title={ '' }
          aria-disabled={ item.inactive }
          data-command-key={ itemKey }
          aria-haspopup={ hasSubmenu(item) }
          aria-expanded={ hasSubmenu(item) ? expandedMenuItemKey === item.key : undefined }
          role='menuitem'
          aria-label={ ariaLabel }
          aria-setsize={ setSize }
          aria-posinset={ posInSet }
        >
          { (hasIcon) ? this._renderIcon(item) : (null) }
          { isNameVisible && (
            <span
              className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
            >
              { item.name }
            </span>
          ) }
          { hasSubmenu(item) ? (
            <Icon className={ css('ms-CommandBarItem-chevronDown', styles.itemChevronDown) } iconName='ChevronDown' />
          ) : (null) }
        </button>
      );
    } else {
      // Allow the disabled property on div elements for commandbar
      command = (
        <div
          { ...getNativeProps(item, divProperties.concat(['disabled'])) }
          id={ this._id + item.key }
          className={ className }
          title={ '' }
          aria-disabled={ item.inactive }
          data-command-key={ itemKey }
          aria-haspopup={ hasSubmenu(item) }
          aria-label={ ariaLabel }
          aria-setsize={ setSize }
          aria-posinset={ posInSet }
        >
          { (hasIcon) ? this._renderIcon(item) : (null) }
          { (isNameVisible) && (
            <span
              className={ css('ms-CommandBarItem-commandText', styles.itemCommandText) }
              aria-hidden='true'
              role='presentation'
            >
              { item.name }
            </span>
          ) }
        </div>
      );
    }

    if (item.iconOnly && item.name) {
      command = (
        <TooltipHost content={ item.name }>
          { command }
        </TooltipHost>
      );
    } else if (tooltipContent) {
      command = (
        <TooltipHost content={ tooltipContent }>
          { command }
        </TooltipHost>
      );
    }

    return (
      <div className={ css('ms-CommandBarItem', styles.item, item.className) } key={ itemKey } ref={ itemKey }>
        { command }
      </div>
    );
  }

  private _renderIcon(item: IContextualMenuItem): JSX.Element {
    // Only present to allow continued use of item.icon which is deprecated.
    const iconProps: IIconProps = item.iconProps ? item.iconProps : {
      iconName: item.icon
    };
    // Use the default icon color for the known icon names
    const iconColorClassName = iconProps.iconName === 'None' ? '' : ('ms-CommandBarItem-iconColor ' + styles.itemIconColor);
    const iconClassName = css('ms-CommandBarItem-icon', styles.itemIcon, iconColorClassName, iconProps.className);

    return <Icon { ...iconProps } className={ iconClassName } />;
  }

  private _asyncMeasure(): void {
    this._async.requestAnimationFrame(() => {
      this._updateItemMeasurements();
      this._updateRenderedItems();
    });
  }

  private _updateItemMeasurements(): void {
    // the generated width for overflow is 35 in chrome, 38 in IE, but the actual value is 41.5
    if (this._overflow.current || (this.props.overflowItems && this.props.overflowItems.length)) {
      this._overflowWidth = OVERFLOW_WIDTH;
    } else {
      this._overflowWidth = 0;
    }

    if (!this._commandItemWidths) {
      this._commandItemWidths = {};
    }

    for (let i = 0; i < this.props.items.length; i++) {
      const item = this.props.items[i];

      if (!this._commandItemWidths[item.key]) {
        const el = this.refs[item.key] as HTMLElement;

        if (el) {
          this._commandItemWidths[item.key] = el.getBoundingClientRect().width;
        }
      }
    }
  }

  private _updateRenderedItems(): void {
    const { items, overflowItems } = this.props;
    const commandSurface = this._commandSurface.current;
    const farCommandSurface = this._farCommandSurface.current;
    const commandBarRegion = this._commandBarRegion.current;
    const searchSurface = this._searchSurface.current;
    const renderedItems = [...items];
    let renderedOverflowItems = overflowItems;
    let consumedWidth = 0;
    const isOverflowVisible = overflowItems && overflowItems.length;

    if (!commandSurface || !commandBarRegion) {
      return;
    }

    const style = window.getComputedStyle(commandSurface);
    let availableWidth = commandBarRegion.clientWidth - parseInt(style.marginLeft!, 10) - parseInt(style.marginRight!, 10);
    if (searchSurface) {
      availableWidth -= searchSurface.getBoundingClientRect().width;
    }
    if (farCommandSurface) {
      availableWidth -= farCommandSurface.getBoundingClientRect().width;
    }

    if (isOverflowVisible) {
      availableWidth -= this._overflowWidth;
    }

    if (!this._commandItemWidths) {
      this._asyncMeasure();
      return;
    }

    for (let i = 0; i < renderedItems.length; i++) {
      const item = renderedItems[i];
      const itemWidth = this._commandItemWidths[item.key];

      if ((consumedWidth + itemWidth) >= availableWidth) {
        if (i > 0 && !isOverflowVisible && (availableWidth - consumedWidth) < OVERFLOW_WIDTH) {
          i--;
        }

        renderedOverflowItems = renderedItems.splice(i).concat(overflowItems!);
        break;
      } else {
        consumedWidth += itemWidth;
      }

    }

    const renderedContextualMenuProps = this._getContextualMenuPropsAfterUpdate(
      renderedItems.concat(this.state.renderedFarItems!),
      renderedOverflowItems!);

    this.setState({
      renderedItems: renderedItems,
      renderedOverflowItems: renderedOverflowItems,
      expandedMenuItemKey: renderedContextualMenuProps ? this.state.expandedMenuItemKey : undefined,
      contextualMenuProps: renderedContextualMenuProps!,
      contextualMenuTarget: renderedContextualMenuProps ? this.state.contextualMenuTarget : undefined
    });
  }

  private _onItemClick(item: IContextualMenuItem): (ev: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void {
    return (ev: React.MouseEvent<HTMLButtonElement>): void => {
      if (item.inactive) {
        return;
      }

      if (item.key === this.state.expandedMenuItemKey || !hasSubmenu(item)) {
        this._onContextMenuDismiss();
      } else {
        this.setState({
          expandedMenuId: ev.currentTarget.id,
          expandedMenuItemKey: item.key,
          contextualMenuProps: this._getContextualMenuPropsFromItem(item),
          contextualMenuTarget: ev.currentTarget
        });
      }
      if (item.onClick) {
        item.onClick(ev, item);
      }
    };
  }

  private _onOverflowClick = (ev: React.MouseEvent<HTMLButtonElement>) => {
    if (this.state.expandedMenuItemKey === OVERFLOW_KEY) {
      this._onContextMenuDismiss();
    } else {
      this.setState({
        expandedMenuId: ev.currentTarget.id,
        expandedMenuItemKey: OVERFLOW_KEY,
        contextualMenuProps: { items: this.state.renderedOverflowItems! },
        contextualMenuTarget: ev.currentTarget
      });
    }
  }

  private _onContextMenuDismiss = (ev?: any) => {
    const { current: commandSurface } = this._commandSurface;

    if (!ev || !ev.relatedTarget || commandSurface && !commandSurface.contains(ev.relatedTarget as HTMLElement)) {
      const { contextualMenuProps } = this.state;

      if (contextualMenuProps && contextualMenuProps.onDismiss) {
        this.state.contextualMenuProps!.onDismiss!(ev);
      }

      this.setState({
        expandedMenuItemKey: undefined,
        contextualMenuProps: undefined,
        contextualMenuTarget: undefined
      });
    } else {
      ev.stopPropagation();
      ev.preventDefault();
    }
  }

  private _getStateFromProps(nextProps: ICommandBarProps): ICommandBarState {
    return {
      renderedItems: nextProps.items || [],
      contextualMenuProps: this._getContextualMenuPropsAfterUpdate(
        nextProps.items.concat(nextProps.farItems!),
        nextProps.overflowItems!)!,
      renderedFarItems: nextProps.farItems || undefined
    };
  }

  private _getContextualMenuPropsAfterUpdate(renderedItems: IContextualMenuItem[], overflowItems: IContextualMenuItem[]): IContextualMenuProps | undefined {
    if (this.state && this.state.expandedMenuItemKey) {
      if (this.state.expandedMenuItemKey === OVERFLOW_KEY) {
        // Keep the overflow menu open
        return { items: overflowItems };
      } else {
        // Find the currently open key in the new props
        const matchingItem = renderedItems.filter(item => item.key === this.state.expandedMenuItemKey);

        if (matchingItem.length === 1) {
          return this._getContextualMenuPropsFromItem(matchingItem[0]);
        }
      }
    }

    return;
  }

  private _getContextualMenuPropsFromItem(item: IContextualMenuItem): IContextualMenuProps | undefined {
    return item.subMenuProps || (item.items && { items: item.items });
  }
}
