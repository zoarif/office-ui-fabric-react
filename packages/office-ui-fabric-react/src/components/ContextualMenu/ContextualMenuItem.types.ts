import { IContextualMenuItem } from './ContextualMenu.types';
import { IMenuItemClassNames } from './ContextualMenu.classNames';

export interface IContextualMenuItemProps
  extends React.HTMLAttributes<IContextualMenuItemProps> {

  /**
   * The item to display
   */
  item: IContextualMenuItem;

  /**
   * Classnames for different aspects of a menu item
   */
  classNames: IMenuItemClassNames;

  /**
   * Index of the item
   */
  index: number;

  /**
   * If this item has icons
   */
  hasIcons: boolean | undefined;

  /**
   * Click handler for the checkmark
   */
  onCheckmarkClick?: ((item: IContextualMenuItem, ev: React.MouseEvent<HTMLElement>) => void);
}
