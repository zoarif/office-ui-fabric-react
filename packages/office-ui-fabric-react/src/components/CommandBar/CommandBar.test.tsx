import * as React from 'react';

import * as ReactDOM from 'react-dom';
import * as ReactTestUtils from 'react-dom/test-utils';
import * as renderer from 'react-test-renderer';

import { CommandBar } from './CommandBar';
import { IContextualMenuItem } from '../ContextualMenu/ContextualMenu.types';

describe('CommandBar', () => {

  afterEach(() => {
    for (let i = 0; i < document.body.children.length; i++) {
      if (document.body.children[i].tagName === 'DIV') {
        document.body.removeChild(document.body.children[i]);
        i--;
      }
    }
  });

  it('renders CommandBar correctly', () => {
    const items: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        subMenuProps: {
          items: [
            {
              name: 'SubmenuText 1',
              key: 'SubmenuKey1',
              className: 'SubMenuClass'
            }
          ]
        }
      },
    ];
    const component = renderer.create(
      <CommandBar items={ items } />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds the correct aria-setsize and -posinset attributes to the command bar items.', () => {
    const items: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        className: 'item1',
        subMenuProps: {
          items: [
            {
              name: 'SubmenuText 1',
              key: 'SubmenuKey1',
              className: 'SubMenuClass'
            }
          ]
        }
      },
      {
        name: 'TestText 2',
        key: 'TestKey2',
        className: 'item2',
      },
      {
        name: 'TestText 3',
        key: 'TestKey3',
        className: 'item3',
      },
    ];

    const renderedContent = ReactTestUtils.renderIntoDocument<CommandBar>(
      <CommandBar
        items={ items }
      />
    ) as React.Component<CommandBar, {}>;
    document.body.appendChild(ReactDOM.findDOMNode(renderedContent) as Element);

    const [item1, item2, item3] = ['.item1', '.item2', '.item3'].map(i => document.querySelector(i)!.children[0]);
    expect(item1.getAttribute('aria-setsize')).toBe('3');
    expect(item2.getAttribute('aria-setsize')).toBe('3');
    expect(item3.getAttribute('aria-setsize')).toBe('3');
    expect(item1.getAttribute('aria-posinset')).toBe('1');
    expect(item2.getAttribute('aria-posinset')).toBe('2');
    expect(item3.getAttribute('aria-posinset')).toBe('3');
  });

  it('opens a menu with deprecated IContextualMenuItem.items property', () => {
    const items: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        items: [
          {
            name: 'SubmenuText 1',
            key: 'SubmenuKey1',
            className: 'SubMenuClass'
          }
        ]
      },
    ];

    const renderedContent = ReactTestUtils.renderIntoDocument<CommandBar>(
      <CommandBar
        items={ items }
      />
    ) as React.Component<CommandBar, {}>;
    document.body.appendChild(ReactDOM.findDOMNode(renderedContent) as Element);

    const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;
    ReactTestUtils.Simulate.click(menuItem);

    expect(document.querySelector('.SubMenuClass')).toBeDefined();
  });

  it('opens a menu with IContextualMenuItem.subMenuProps.items property', () => {
    const items: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        subMenuProps: {
          items: [
            {
              name: 'SubmenuText 1',
              key: 'SubmenuKey1',
              className: 'SubMenuClass'
            }
          ]
        }
      },
    ];

    const renderedContent = ReactTestUtils.renderIntoDocument<CommandBar>(
      <CommandBar
        items={ items }
      />
    ) as React.Component<CommandBar, {}>;
    document.body.appendChild(ReactDOM.findDOMNode(renderedContent) as Element);

    const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;
    ReactTestUtils.Simulate.click(menuItem);

    expect(document.querySelector('.SubMenuClass')).toBeDefined();
  });

  it('keeps menu open after update if item is still present', () => {
    const renderContainer = document.createElement('div');
    document.body.appendChild(renderContainer);

    try {
      const items: IContextualMenuItem[] = [
        {
          name: 'TestText 1',
          key: 'TestKey1',
          subMenuProps: {
            items: [
              {
                name: 'SubmenuText 1',
                key: 'SubmenuKey1',
                className: 'SubMenuClass'
              }
            ]
          }
        },
      ];

      let renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;
      ReactTestUtils.Simulate.click(menuItem);

      // Make sure the menu is open before the re-render
      expect(document.querySelector('.SubMenuClass')).toBeDefined();

      // Update the props, and re-render
      items.push({
        name: 'Test Key 2',
        key: 'TestKey2'
      });

      renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      // Make sure the menu is still open after the re-render
      expect(document.querySelector('.SubMenuClass')).toBeDefined();
    } finally {
      ReactDOM.unmountComponentAtNode(renderContainer);
      document.body.removeChild(renderContainer);
    }
  });

  it('closes menu after update if item is not longer present', () => {
    const renderContainer = document.createElement('div');
    document.body.appendChild(renderContainer);

    try {
      let items: IContextualMenuItem[] = [
        {
          name: 'TestText 1',
          key: 'TestKey1',
          subMenuProps: {
            items: [
              {
                name: 'SubmenuText 1',
                key: 'SubmenuKey1',
                className: 'SubMenuClass'
              }
            ]
          }
        },
      ];

      let renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;
      ReactTestUtils.Simulate.click(menuItem);

      // Make sure the menu is open before the re-render
      expect(document.querySelector('.SubMenuClass')).toBeDefined();

      // Update the props, and re-render
      items = [{
        name: 'Test Key 2',
        key: 'TestKey2'
      }];

      renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      // Make sure the menu is still open after the re-render
      expect(document.querySelector('.SubMenuClass')).toBeNull();
    } finally {
      ReactDOM.unmountComponentAtNode(renderContainer);
      document.body.removeChild(renderContainer);
    }
  });

  it('updates menu after update if item is still present', () => {
    const renderContainer = document.createElement('div');
    document.body.appendChild(renderContainer);

    try {
      const items: IContextualMenuItem[] = [
        {
          name: 'TestText 1',
          key: 'TestKey1',
          subMenuProps: {
            items: [
              {
                name: 'SubmenuText 1',
                key: 'SubmenuKey1',
                className: 'SubMenuClass'
              }
            ]
          }
        },
      ];

      let renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;
      ReactTestUtils.Simulate.click(menuItem);

      // Make sure the menu is open before the re-render
      expect(document.querySelector('.SubMenuClass')).toBeDefined();

      // Update the props, and re-render
      items[0].subMenuProps!.items[0].className = 'SubMenuClassUpdate';

      renderedContent = ReactDOM.render(
        <CommandBar
          items={ items }
        />,
        renderContainer
      ) as React.Component<CommandBar, {}>;

      // Make sure the menu is still open after the re-render
      expect(document.querySelector('.SubMenuClass')).toBeNull();
      expect(document.querySelector('.SubMenuClassUpdate')).toBeDefined();
    } finally {
      ReactDOM.unmountComponentAtNode(renderContainer);
      document.body.removeChild(renderContainer);
    }
  });

  it('Command bar item shows chevron even if submenu has no items', () => {
    const menuWithNoSubmenu: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        subMenuProps: {
          items: []
        }
      },
    ];

    const renderedContent = ReactTestUtils.renderIntoDocument<CommandBar>(
      <CommandBar
        items={ menuWithNoSubmenu }
      />
    ) as React.Component<CommandBar, {}>;
    const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;

    expect(menuItem.querySelector('.ms-CommandBarItem-chevronDown')).not.toEqual(null);
  });

  it('Command bar item onMenuOpened is called even if submenu has no items', () => {
    let subMenuOpened = false;

    const onSubMenuOpened = (): void => {
      subMenuOpened = true;
    };

    const menuWithEmptySubMenu: IContextualMenuItem[] = [
      {
        name: 'TestText 1',
        key: 'TestKey1',
        subMenuProps: {
          items: [],
          onMenuOpened: onSubMenuOpened
        }
      }
    ];

    const renderedContent = ReactTestUtils.renderIntoDocument<CommandBar>(
      <CommandBar
        items={ menuWithEmptySubMenu }
      />
    ) as React.Component<CommandBar, {}>;

    const menuItem = (ReactDOM.findDOMNode(renderedContent) as HTMLElement).querySelector('button') as HTMLButtonElement;

    ReactTestUtils.Simulate.click(menuItem);

    expect(subMenuOpened).toEqual(true);
  });

});