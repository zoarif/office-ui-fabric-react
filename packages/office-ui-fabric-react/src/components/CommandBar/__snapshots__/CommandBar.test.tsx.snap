// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandBar renders CommandBar correctly 1`] = `
<div
  className="ms-CommandBar"
>
  <div
    aria-describedby={undefined}
    aria-labelledby={undefined}
    className="ms-FocusZone"
    data-focuszone-id="FocusZone1"
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDownCapture={[Function]}
    role="menubar"
  >
    <div
      className="ms-CommandBar-primaryCommands"
    >
      <div
        className="ms-CommandBarItem"
      >
        <button
          aria-disabled={undefined}
          aria-expanded={false}
          aria-haspopup={true}
          aria-label={undefined}
          aria-posinset={1}
          aria-setsize={1}
          className="ms-CommandBarItem-link undefined"
          data-command-key="TestKey1"
          id="CommandBar0TestKey1"
          name="TestText 1"
          onClick={[Function]}
          role="menuitem"
          title=""
        >
          <span
            className="ms-CommandBarItem-commandText"
          >
            TestText 1
          </span>
          <i
            aria-hidden={true}
            className=
                ms-CommandBarItem-chevronDown
                {
                  display: inline-block;
                }
            data-icon-name="ChevronDown"
            role="presentation"
          />
        </button>
      </div>
    </div>
    <div
      className="ms-CommandBar-sideCommands"
    />
  </div>
</div>
`;
